File: 3329-find-the-length-of-the-longest-common-prefix/3329-find-the-length-of-the-longest-common-prefix.cpp
#pragma GCC optimize(O3, "unroll-loops")

class Solution {
public:
    int longestCommonPrefix(vector<int>& arr1, vector<int>& arr2) {
        unordered_set<int> pref;
        for (int val: arr1) {
            while (!pref.count(val) && val) {
                pref.insert(val);
                val /= 10;
            }
        }

        int maxPref = 0;

        for (int val: arr2) {
            while (!pref.count(val) && val) val /= 10;
            if (val) maxPref = max(maxPref, static_cast<int>(log10(val) + 1));
        }

        return maxPref;
    }
};

auto init = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 'c';
}();

File: 3329-find-the-length-of-the-longest-common-prefix/3329-find-the-length-of-the-longest-common-prefix.kt
class Solution {
    fun longestCommonPrefix(arr1: IntArray, arr2: IntArray): Int {
        val pref: HashSet<Int> = HashSet();
        var t: Int;
        for (v in arr1) {
            t = v;
            while (!pref.contains(t) && t > 0) {
                pref.add(t);
                t /= 10;
            }
        }

        var maxPref: Int = 0;

        for (v in arr2) {
            t = v;
            while (!pref.contains(t) && t > 0) t /= 10;
            if (t > 0) maxPref = max(maxPref, (log10(t.toFloat()) + 1).toInt());
        }

        return maxPref;
    }
}